### PR TITLE
[2021.2][SG] Blackboard Categories 

### PR DESCRIPTION
**Pull request by Nightmask3**
---
### Purpose of this PR

This PR adds the ability to create categories in the blackboard, that can then be used to organize your blackboard items, both within the shader graph, and in the Material Inspector, for materials created based on a shader graph.

This branch is a staging branch, where each individual feature that composes the categories as a whole was developed on a separate feature branch that was reviewed before being merged into the staging branch. 

There has also been QA testing of the features as they were progressively added, so some of this code has had QA coverage prior to this PR being created. 

This feature can be broken down into 2 parts, at a high level:
- The code particular to the categories UI stuff and data management in the shader graph
- The code that handles the material inspector representation of the shader graph categories and properties/keywords

This PR also adds code that changes how shader graph properties are rendered across URP and HDRP for shader graph based materials, inserting a custom property drawer (see `ShaderGraphPropertyDrawers`) that overrides the default material inspector handling of shader graph properties, allowing for a custom UI field representation of:
- Vector2/Vector3/Vector4s, where before they were all represented by a single vec4 field 
- Virtual Textures, which now indent the layers that compose the virtual texture property into a foldout that can be collapsed and expanded
-  The categories and the properties/keywords contained within them.

The categories working in a URP project:

https://user-images.githubusercontent.com/13033542/117181039-ad076580-ad89-11eb-8133-c2bd27cfb651.mp4

The categories working in a HDRP project

https://user-images.githubusercontent.com/13033542/117186031-4e44ea80-ad8f-11eb-92e7-c7493a3f519d.mp4

Link to the QA test doc by Landon: https://docs.google.com/document/d/1iwLU5TM3WbQCKRzm1iSMQAlO1SjVkZv74nK5JlLLvJQ/edit

Link to the spec doc **(which has details of the expected user interactions, might be useful for testing)**: https://docs.google.com/document/d/1l87jXXVkVTgDFA-OnKK5Je4BPw0_C5QcCL8uaTwYyaM/edit#heading=h.9a8vlbymtc9e

Link to the Jira card: https://jira.unity3d.com/browse/GSG-258

---
### Testing status

- QA testing performed by Landon over the course of development, as well as on this PR specifically 
- Tech-Artist testing being performed by some volunteers organized by Cici and Tracy, details of which can be found in this Slack channel: https://unity.slack.com/archives/C01T63JMB2T
- [URP_Foundation Win Vulkan playmode](https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Fblackboard-categories-staging/.yamato%252Furp_foundation-win-vulkan.yml%2523URP_Foundation_Win_Vulkan_playmode_cache_mono_Linear_trunk/6767484/job) fails UniversalGraphicsTests.245_Normal_Reconstruction.  This matches master.
- [HDRP OSX Metal playmode](https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Fblackboard-categories-staging/.yamato%252Fhdrp-osx-metal.yml%2523HDRP_OSX_Metal_playmode_cache_mono_Linear_trunk/6782699/job/(log:Execution)) asset import errors with IES files matches master
- [URP_Lighting on iPhone](https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%2Fblackboard-categories-staging/.yamato%2Furp_lighting-iphone-metal.yml%23URP_Lighting_iPhone_Metal_Standalone_cache_il2cpp_Linear_trunk/6782681/job/(log:Execution)) UniversalGraphicsTests.017_Lighting_Scene_DirectionalBaked fails.  This matches master.
- [URP_Terrain on iPhone](https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Fblackboard-categories-staging/.yamato%252Furp_terrain-iphone-metal.yml%2523Build_URP_Terrain_iPhone_Metal_Standalone_cache_il2cpp_Linear_trunk/6767557/job/(log:Execution)) non-test failure.  This matches master.
- Other failures were also observed in master on recent PRs, so this seems in parity with current master.
---
### Comments to reviewers

Things to do still:

- [x] Address review comments
- [x] Fix QA reported bugs
- [x] Add tests for the various category based actions
   - Tests are added for adding/deleting/duplicating/renaming/expand-collapse categories
   - Currently don't have tests for moving categories/adding items to categories code paths, as both of these involve drag/drop and as the assembly that allows for the automation of drag/drop stuff hides all its functionality in an internal class, seems like we're out of luck until we can promote those functions to public in the future or expose internals to SG assembly for testing.

Will do in a follow up PR: 
- Fix UI layout of Shader Graph based material properties and categories in Material Inspector to make it flow better and look better

### CODE REVIEW

**Comment by Nightmask3** [comment]
> It appears that you made a non-draft PR!
> Please convert your PR to draft (button on the right side of the page)
> and cancel any jobs that started on Yamato.
> See the PR template for more information.
> Thank you!

Given that we need to trigger Yamato jobs manually now anyway, making it a draft PR seemed pointless, correct me if I'm wrong though

File: com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs [hunk 1]
```diff
@@ -5,14 +5,53 @@
...
+        public ShaderKeyword builtInKeyword;
+        public string deprecatedPropertyName;
+        public int version;
+    }
+    class BlackboardShaderInputFactory
+    {
+        static public ShaderInput GetShaderInput(BlackboardShaderInputOrder order)
+        {
+            ShaderInput output;
+            if (order.isKeyword)
```
Comment (russoh) [inline com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs:27]: Does this need handling for dropdowns?
else if (order.isDropdown)...

File: com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardCategoryController.cs [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+                    break;
+
+                case DeleteCategoryAction deleteCategoryAction:
+                    if (deleteCategoryAction.categoriesToRemoveGuids.Contains(ViewModel.associatedCategoryGuid))
+                    {
+                        this.Destroy();
+                        return;
+                    }
+
+                    // Check if any inputs were added to this category (i.e. if a category merge took place)
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardCategoryController.cs:204]: Can't merge categories anymore so see if this code is still relevant, remove if not

File: com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs (hunk 1 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs:27]: If I drag properties from one subgraph to another, the properties will often not get copied. I'm not sure if this is new but it looks like it only works if you release the mouse button over the top of one of the existing properties?

Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs:27]: Duplicating properties does not appear to work any more. On the blackboard, right-click a properties, Duplicate -> nothing appears to be duplicated.

File: com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs [hunk 3]
```diff
@@ -212,62 +408,48 @@ void InitializeViewModel(bool useDropdowns)
...
                 }
                 else
                 {
-                    ViewModel.builtInKeywordNameToAddActionMap.Add(keyword.displayName,  new AddShaderInputAction() { shaderInputReference = keyword.Copy(), addInputActionType = AddShaderInputAction.AddActionSource.AddMenu });
+                    order.builtInKeyword = (ShaderKeyword)keyword.Copy();
+                    ViewModel.builtInKeywordNameToAddActionMap.Add(keyword.displayName, AddShaderInputAction.AddKeywordAction(order));
                 }
             }
 
             if (useDropdowns)
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs:444]: As @russoh pointed out we'll need to handle the dropdowns same as the others, using `BlackboardShaderInputOrder`, missed it during the merge, good catch

File: com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs (hunk 1 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs:27]: Yup this is missing handling for the dropdowns, missed it during merging in from master cause it "just worked"

Will also look at why they dont get added to selected categories 👍

Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs:27]: Landon caught the breakage of dragging properties/categories between graphs, I didn't think to handle that, will address it in the next iteration

The two categories reversing bug is new, good catch, will look into that

**Comment by LandonTownsendUnity** [comment]
test doc:
https://docs.google.com/document/d/1iwLU5TM3WbQCKRzm1iSMQAlO1SjVkZv74nK5JlLLvJQ/edit?usp=sharing

**Comment by hybridherbst** [comment]
Great that there's some movement in this area!
When you're adding categories like this, will this also come with the ability to add custom attributes/drawers/decorators to properties?

(Disclosure: I'm the dev behind [Shader Graph Markdown](https://openupm.com/packages/com.needle.shadergraph-markdown/) and have worked on something very similar, but backwards compatible until URP/HDRP 7 and with much bigger feature coverage.)

I'd be very happy if Shader Graph gains more built-in customization options and would love if that doesn't fall short of expectations!

**Comment by hybridherbst** [comment]
It looks like the relevant drawers etc of this PR are all internal. How are users intended to build custom shader GUIs that use this new "categories" feature?

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs [hunk 4]
```diff
@@ -17,37 +17,117 @@ public string name
...
+            m_ChildObjectList.Remove(itemToRemove);
+            if (m_ChildObjectIDSet.Contains(itemToRemove.objectId))
+                m_ChildObjectIDSet.Remove(itemToRemove.objectId);
+        }
+
+        public void MoveItemInCategory(ShaderInput itemToMove, int newIndex)
+        {
+            m_ChildObjectList.Remove(itemToMove);
+            m_ChildObjectList.Insert(newIndex,itemToMove);
+            // Recreate the hash-set as the data structure does not allow for moving/inserting elements
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:68]: Same comment as above

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs [hunk 5]
```diff
@@ -17,37 +17,117 @@ public string name
...
+        public void RemoveItemFromCategory(ShaderInput itemToRemove)
+        {
+            m_ChildObjectList.Remove(itemToRemove);
+            if (m_ChildObjectIDSet.Contains(itemToRemove.objectId))
+                m_ChildObjectIDSet.Remove(itemToRemove.objectId);
+        }
+
+        public void MoveItemInCategory(ShaderInput itemToMove, int newIndex)
+        {
+            m_ChildObjectList.Remove(itemToMove);
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:66]: Any reason this can't just call `RemoveItemFromCategory` then `InsertItemIntoCategory`. There's a lot of error checking you do in Insert in particular that isn't done here (null item, -1 index, etc...)

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs [hunk 6]
```diff
@@ -17,37 +17,117 @@ public string name
...
+        }
 
-        HashSet<Guid> m_ChildItemIDSet;
-        // We expose Guids as a HashSet for faster existence checks
-        public HashSet<Guid> childItemIDSet
+        public bool IsNamedCategory()
         {
-            get => m_ChildItemIDSet;
-            set => m_ChildItemIDSet = value;
+            return name != String.Empty;
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:83]: Is null possible here? Would it be better to use `string.IsNullOrEmpty`?

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs [hunk 7]
```diff
@@ -17,37 +17,117 @@ public string name
...
+                {
+                    m_ChildObjectList.Add(childObject);
+                    m_ChildObjectIDSet.Add(childObject.objectId);
+                }
+            }
+        }
+
+        public CategoryData(CategoryData categoryToCopy)
+        {
+            this.name = categoryToCopy.name;
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:125]: Shouldn't category copying also copy the objects?

File: com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboard.cs [hunk 8]
```diff
@@ -175,14 +207,195 @@ public void ShowScrollBoundaryRegions()
...
+                        ++index;
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+            }
+
+            if (index == -1 && childCount >= 2)
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboard.cs:249]: A small comment here might be nice. I can guess that this is handling dragging before/after the entire control, but it's not clear to my why `childCount >= 2` is here.

File: com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs [hunk 9]
```diff
@@ -0,0 +1,623 @@
...
+        internal TextField textField => m_TextField;
+        VisualElement m_RowsContainer;
+        int m_InsertIndex;
+        SGBlackboard blackboard => m_ViewModel.parentView as SGBlackboard;
+
+        public delegate bool CanAcceptDropDelegate(ISelectable selected);
+
+        public CanAcceptDropDelegate canAcceptDrop { get; set; }
+
+        int InsertionIndex(Vector2 pos)
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs:73]: I've seen this logic in a few places, would it be possible to extract to some shared location?

File: com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs [hunk 10]
```diff
@@ -0,0 +1,623 @@
...
+
+        public override VisualElement contentContainer { get { return m_RowsContainer; } }
+
+        public override string title
+        {
+            get => m_TitleLabel.text;
+            set
+            {
+                m_TitleLabel.text = value;
+                if (m_TitleLabel.text == String.Empty)
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs:197]: Should this instead use `string.IsNullOrEmpty`?

File: com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs [hunk 11]
```diff
@@ -0,0 +1,623 @@
...
+            else
+                m_MainContainer.Add(m_RowsContainer);
+
+            var key = $"{blackboard.controller.editorPrefsBaseKey}.{viewDataKey}.{ChangeCategoryIsExpandedAction.kEditorPrefKey}";
+            EditorPrefs.SetBool(key, expand);
+        }
+
+        void OnMouseDownEvent(MouseDownEvent e)
+        {
+            if ((e.clickCount == 2) && e.button == (int)MouseButton.LeftMouse && IsRenamable())
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs:282]: Is this handling double click? A simple comment stating such might be nice

File: com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs [hunk 12]
```diff
@@ -0,0 +1,623 @@
...
+            m_TextField.Q(TextField.textInputUssName).Focus();
+            m_TextField.SelectAll();
+        }
+
+        void OnEditTextFinished()
+        {
+            m_TitleLabel.visible = true;
+            m_TextField.style.display = DisplayStyle.None;
+
+            if (title != m_TextField.text && String.IsNullOrWhiteSpace(m_TextField.text) == false && String.IsNullOrEmpty(m_TextField.text) == false)
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs:303]: `IsNullOrWhiteSpace` should be sufficient here given the MSDN description:
```true if the value parameter is null or Empty, or if value consists exclusively of white-space characters.```

File: com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs [hunk 13]
```diff
@@ -192,13 +372,36 @@ internal SGBlackboard blackboard
             get => m_Blackboard;
             private set => m_Blackboard = value;
         }
+        public string GetFirstSelectedCategoryGuid()
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs:375]: This looks like it returns `String.Empty` if multiple categories are selected. Why is this the case?

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs [hunk 14]
```diff
@@ -841,25 +841,38 @@ void ConvertToInlineNode(DropdownMenuAction action)
 
         void DuplicateSelection()
         {
-            graph.owner.RegisterCompleteObjectUndo("Duplicate Blackboard Property");
+            graph.owner.RegisterCompleteObjectUndo("Duplicate Blackboard Selection");
 
             List<ShaderInput> selectedProperties = new List<ShaderInput>();
+            List<CategoryData> selectedCategories = new List<CategoryData>();
+
+            foreach (var selectable in selection.ToList())
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs:849]: Why `ToList()`? Can't you just enumerate the selection?

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs [hunk 15]
```diff
@@ -878,9 +891,19 @@ string SerializeGraphElementsImplementation(IEnumerable<GraphElement> elements)
             var groups = elements.OfType<ShaderGroup>().Select(x => x.userData);
             var nodes = elements.OfType<IShaderNodeView>().Select(x => x.node).Where(x => x.canCopyNode);
             var edges = elements.OfType<Edge>().Select(x => (Graphing.Edge)x.userData);
-            var inputs = selection.OfType<BlackboardPropertyView>().Select(x => x.userData as ShaderInput).ToList();
             var notes = elements.OfType<StickyNote>().Select(x => x.userData);
 
+            var categories = new List<CategoryData>();
+            foreach (var selectable in selection.ToList())
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs:897]: Why `ToList()`? Can't you just enumerate the selection?

File: com.unity.shadergraph/Editor/Drawing/MaterialEditor/ShaderGraphPropertyDrawers.cs [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+    {
+        static Dictionary<MinimalCategoryData.GraphInputData, bool> s_CompoundPropertyFoldoutStates = new Dictionary<MinimalCategoryData.GraphInputData, bool>();
+
+        public static void DrawShaderGraphGUI(MaterialEditor materialEditor, MaterialProperty[] properties)
+        {
+            Material m = materialEditor.target as Material;
+            Shader s = m.shader;
+            string path = AssetDatabase.GetAssetPath(s);
+            ShaderGraphMetadata metadata = null;
+            foreach (var obj in AssetDatabase.LoadAllAssetsAtPath(path))
```
[the remark this answers is not in the document]
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/MaterialEditor/ShaderGraphPropertyDrawers.cs:22]: Hey Felipe, good question, so I threw in a breakpoint to test this, and it looks like this gets called whenever the Material Inspector triggers its own internal update cycle, which appears to be when:
- The user clicks on a Material asset derived from a shader graph (triggering the material inspector)
- The user moves their mouse across an input field 
- Whenever a shader graph is saved, changing the Material asset on disk (requiring a refresh of the inspector)

**Comment by marctem** [comment]
> It looks like the relevant drawers etc of this PR are all internal. How are users intended to build custom shader GUIs that use this new "categories" feature?

Hey, @hybridherbst, absolutely fair question. At this time we're not providing helpers to enable categories in custom shader guis because we'll be refactoring them in the next version. We are going that direction but for now it will require a completely custom gui that explicitly creates the categories in code.

**Comment by Nightmask3** [comment]
> > It looks like the relevant drawers etc of this PR are all internal. How are users intended to build custom shader GUIs that use this new "categories" feature?
> 
> Hey, @hybridherbst, absolutely fair question. At this time we're not providing helpers to enable categories in custom shader guis because we'll be refactoring them in the next version. We are going that direction but for now it will require a completely custom gui that explicitly creates the categories in code.

Towards this end, something that could help is this snippet of code provided by the person who authored the material inspector integration of the blackboard categories:

```
public static IEnumerator<(string categoryName, string propertyReferenceName)> GetCategoryData(Shader shaderGraphShader)
        {
            string path = AssetDatabase.GetAssetPath(shaderGraphShader);
            ShaderGraphMetadata metadata = null;
            foreach (var obj in AssetDatabase.LoadAllAssetsAtPath(path))
            {
                if (obj is ShaderGraphMetadata meta)
                {
                    metadata = meta;
                    break;
                }
            }
            Debug.Assert(metadata != null, "Cannot get category data on a non-ShaderGraph material");

            foreach(MinimalCategoryData categoryData in metadata.categoryDatas)
            {
                foreach(var propertyData in categoryData.propertyDatas)
                {
                    yield return (categoryData.categoryName, propertyData.referenceName);
                }
            }
        }
```

`MinimalCategoryData` is not public, but its a small data only struct that you can grab the type of and then use to reflect the category data in your custom shader GUI class as needed. 

The code in `ShaderGraphPropertyDrawers` provides a reference as to how we create the categories, but that's very basic IMGUI code for the foldouts and so forth, nothing special there.

File: com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs (hunk 1 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs:28]: These bugs are handled now, and have taken into account dropdowns in this segment of the code

File: com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardCategoryController.cs (hunk 2 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardCategoryController.cs:196]: removing code, can't have items added to a category on a category delete anymore

File: com.unity.shadergraph/Editor/Drawing/MaterialEditor/ShaderGraphPropertyDrawers.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.shadergraph/Editor/Drawing/MaterialEditor/ShaderGraphPropertyDrawers.cs:22]: I'm afraid this might cause editor to hang/hiccups for large projects. Imo you should not do this in the situations above. 
I won't block this PR due to that but this needs more testing/fixing.

[the remark this answers is not in the document]
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/MaterialEditor/ShaderGraphPropertyDrawers.cs:22]: Hey Felipe, this is a good point, though the actual data that we are loading is fairly light, and is not the entire graph data itself as opposed to a slimmed down version that contains the minimal data needed to draw the categories.

I'm asking @LandonTownsendUnity to test out the categories in a large URP project to see if this has any visible impact to the UX or otherwise slows down the project.

File: com.unity.shadergraph/Editor/Data/Graphs/DataStore.cs [hunk 17]
```diff
@@ -20,26 +20,26 @@ internal DataStore(Action<T, ActionType> reducer, T initialState)
...
+            //try
+            //{
                 m_Reducer(State, action);
-            }
-            catch (Exception exception)
-            {
-                Debug.Log("Ran into exception of type: " + exception + " while dispatching action of type: " + action + "to DataStore.");
-                return;
-            }
+            //}
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Graphs/DataStore.cs:26]: are these supposed to be here?  If not, delete the comments.  :)

**Review by cdxntchou — APPROVED** [review]
Found a few GC allocs that can be removed, but that can be done in a follow up.

Looks good!

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs [hunk 18]
```diff
@@ -17,37 +17,117 @@ public string name
...
+
+            if (insertionIndex == -1)
+            {
+                m_ChildObjectList.Add(itemToAdd);
+                m_ChildObjectIDSet.Add(itemToAdd.objectId);
+            }
+            else
+            {
+                m_ChildObjectList.Insert(insertionIndex, itemToAdd);
+                // Recreate the hash-set as the data structure does not allow for moving/inserting elements
```
[the remark this answers is not in the document]
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:48]: From the MSDN docs: 
"The HashSet<T> class provides high-performance set operations. A set is a collection that contains no duplicate elements, and whose elements are in no particular order."

So yeah these are all unnecessary! Will remove 👍

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs (hunk 4 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:68]: Changed it now, left a comment to clarify intent

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs [hunk 19]
```diff
@@ -17,37 +17,117 @@ public string name
...
+                }
+            }
         }
 
-        // We store Guids as a list out to the graph asset
-        [SerializeField]
-        List<Guid> m_ChildItemIDList;
+        public void RemoveItemFromCategory(ShaderInput itemToRemove)
+        {
+            m_ChildObjectList.Remove(itemToRemove);
```
[the remark this answers is not in the document]
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:59]: It would be! Though it occurs to me that the calling code and this code within the function don't maintain any consistency between who is supposed to check for the existence of the item to remove, with some calling locations checking for existence before calling, and some not checking

So it makes sense to me to get rid of the external existence checks and contain that to a single check here so callers can safely call this function, and do so as you say, with both remove calls behind the if check 👍

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs (hunk 5 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:66]: Though I guess one thing that not explicitly calling those existing functions does avoid is having to update the HashSet by adding and removing an item, I think correct-ness trumps performance concerns here

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs (hunk 6 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:83]: The category names are initialized to `String.Empty` at the start so I believe nulls are not possible here, which would also be an assumption I believe a lot of things in the code depend on

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs (hunk 7 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:125]:
```
        public CategoryData CopyCategory(CategoryData categoryToCopy)
        {
            var copiedCategory = new CategoryData(categoryToCopy);
            AddCategory(copiedCategory);
            // Whenever a category is copied, also copy over all the inputs within that category
            foreach (var childInputToCopy in categoryToCopy.Children)
            {
                var newShaderInput = AddCopyOfShaderInput(childInputToCopy);
                copiedCategory.InsertItemIntoCategory(newShaderInput);
            }

            return copiedCategory;
        }
```

But looking at that function again, seeing why I chose to implement it like that, which is that we need to create a copy of the input first, and `AddCopyOfShaderInput` handles a decent amount of stuff that I didn't want to mess with or move from `GraphData`.

File: com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboard.cs (hunk 8 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboard.cs:256]: In any case, clarified the intent with a comment

File: com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs (hunk 9 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs:73]: Extracted the core of the InsertionIndex function into `BlackboardUtils.GetInsertionIndex` that both `SGBlackboard` and `SGBlackboardCategory` call now 👍

File: com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs (hunk 11 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs:282]: will add 👍

File: com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs (hunk 12 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs:303]: sweet will remove that last bit then 👍

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs (hunk 14 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs:849]: Doing it directly without having a `ToList()` is a problem further in the loop, because we're also trying to modify the selection by removing items from it

Not having the `ToList()`, causes an `InvalidOperationException`

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs (hunk 15 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs:897]: Removed the `ToList` here was unnecessary 👍

File: com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs (hunk 14 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Views/MaterialGraphView.cs:849]: But, Chris suggested a better way to do a list enumeration while removing items below in one of his comments, and I'm going to run with that using the actual object indices and `RemoveAll` instead to avoid the list GC here

File: com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs [hunk 20]
```diff
@@ -230,6 +230,81 @@ public override void OnImportAsset(AssetImportContext ctx)
...
+                    MinimalCategoryData.GraphInputData propData;
+                    // Only write out data for exposed blackboard items
+                    if (input.isExposed == false)
+                        continue;
+
+                    // VTs are treated differently
+                    if (input is VirtualTextureShaderProperty virtualTextureShaderProperty)
+                    {
+                        propData = MinimalCategoryData.ProcessVirtualTextureProperty(virtualTextureShaderProperty);
+                        foreach (var inputData in inputInspectorDataList.ToList())
```
[the remark this answers is not in the document]
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs:278]: Changed it to use RemoveAll instead

File: com.unity.render-pipelines.universal/Editor/ShaderGUI/BaseShaderGUI.cs [hunk 21]
```diff
@@ -259,16 +260,7 @@ internal void DrawShaderGraphProperties(Material material, IEnumerable<MaterialP
...
-            {
-                if ((prop.flags & (MaterialProperty.PropFlags.HideInInspector | MaterialProperty.PropFlags.PerRendererData)) != 0)
-                    continue;
-
-                float h = materialEditor.GetPropertyHeight(prop, prop.displayName);
-                Rect r = EditorGUILayout.GetControlRect(true, h, EditorStyles.layerMaskField);
-
-                materialEditor.ShaderProperty(r, prop, prop.displayName);
-            }
+            ShaderGraphPropertyDrawers.DrawShaderGraphGUI(materialEditor, properties.ToArray());
```
[the remark this answers is not in the document]
Comment (Nightmask3) [reply com.unity.render-pipelines.universal/Editor/ShaderGUI/BaseShaderGUI.cs:263]: Can do!

File: com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs (hunk 5 above)
Comment (joshua-davis) [reply com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs:66]: You could maybe refactor the validation of the index etc... to a helper, but I think that removing+inserting into a hashset is so trivial it shouldn't matter

File: com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboard.cs (hunk 8 above)
Comment (joshua-davis) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboard.cs:240]: I would expect those to both work if the childCount is 1 as both are valid locations. Could that be an edge case worth looking into?

**Review by VitaSkruibyte — APPROVED** [review]
Looks good! @Nightmask3 thanks for the feature! Here's the [test doc](https://docs.google.com/document/d/1Ww1WhqrmgLuH0HXxPGGpsAWkdC2-WOXVX7-7o3tfess/edit?usp=sharing)

**Review by LandonTownsendUnity — APPROVED** [review]
Did another quick test after the fixes to make sure everything works. Approving.

**Review by marctem — APPROVED** [review]
Looks good overall!
